File: app/src/main/java/com/mythicalcreaturesoftware/splash/tasks/ImageResizeTask.java
package com.mythicalcreaturesoftware.splash.tasks;

import com.mythicalcreaturesoftware.splash.utils.MathHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;

public class ImageResizeTask implements Runnable {

    private static Logger logger = LoggerFactory.getLogger(ImageResizeTask.class);

    private String inputImagePath;
    private String outputImagePath;
    private double percent;

    public ImageResizeTask(String inputImagePath, String outputImagePath, double percent) {
        this.inputImagePath = inputImagePath;
        this.outputImagePath = outputImagePath;
        this.percent = percent;
    }

    @Override
    public void run() {
        try {
            resize(inputImagePath, outputImagePath, percent);
        } catch (IOException e) {
            logger.error("Error starting thread", e);
        }
    }

    /**
     * Resize an image to a absolute width and height (the image may not be proportional). Both paths should be coming as an encode url, they will be decode during the execution
     * @param inputImagePath Path of the original image
     * @param outputImagePath Path to save the resized image
     * @param scaledWidth absolute width in pixels
     * @param scaledHeight absolute height in pixels
     * @throws IOException IOException
     */
    private void resize(String inputImagePath, String outputImagePath, int scaledWidth, int scaledHeight) throws IOException {
        String cleanedInputImagePath = URLDecoder.decode(inputImagePath, StandardCharsets.UTF_8.toString());
        String cleanedOutputImagePath = URLDecoder.decode(outputImagePath, StandardCharsets.UTF_8.toString());

        // reads input image
        File inputFile = new File(cleanPath(cleanedInputImagePath));
        BufferedImage inputImage = ImageIO.read(inputFile);

        // creates output image
        BufferedImage outputImage = new BufferedImage(scaledWidth, scaledHeight, inputImage.getType());

        // scales the input image to the output image
        Graphics2D g2d = outputImage.createGraphics();
        g2d.drawImage(inputImage, 0, 0, scaledWidth, scaledHeight, null);
        g2d.dispose();

        // extracts extension of output file
        String formatName = cleanedOutputImagePath.substring(cleanedOutputImagePath.lastIndexOf(".") + 1);

        // writes to output file
        File outputFile = new File(cleanPath(cleanedOutputImagePath));
        if (!outputFile.exists()) {
            outputFile.createNewFile();
        }
        ImageIO.write(outputImage, formatName, outputFile);
    }

    /**
     * Resize an image by a percentage of original size (proportional). Both paths should be coming as an encode url, they will be decode during the execution
     * @param inputImagePath Path of the original image
     * @param outputImagePath Path to save the resized image
     * @param percent a double number specifies percentage of the output image. Eg: 10
     * over the input image.
     * @throws IOException IOException
     */
    private void resize(String inputImagePath, String outputImagePath, double percent) throws IOException {
        if (inputImagePath == null || inputImagePath.equals("")) {
            return;
        }
        if (outputImagePath == null || outputImagePath.equals("")) {
            return;
        }

        String cleanedInputImagePath = URLDecoder.decode(inputImagePath, StandardCharsets.UTF_8.toString());
        String cleanedOutputImagePath = URLDecoder.decode(outputImagePath, StandardCharsets.UTF_8.toString());

        File file = new File(cleanPath(cleanedInputImagePath));

        BufferedImage inputImage = ImageIO.read(file);

        if (inputImage != null) {
            int scaledWidth = (int) MathHelper.percentageValue(percent, inputImage.getWidth());
            int scaledHeight = (int) MathHelper.percentageValue(percent, inputImage.getHeight());

            resize(cleanedInputImagePath, cleanedOutputImagePath, scaledWidth, scaledHeight);
        }
    }

    /**
     * Remove the leading "file:" from the indicated path
     * @param path path to clean
     * @return the cleaned path
     */
    private String cleanPath(String path) {
        return path.replace("file:", "");
    }
}

File: reader/src/jvmMain/kotlin/reader/tasks/ImageResizeTask.kt
package reader.tasks

import mu.KotlinLogging
import reader.utils.cleanPath
import reader.utils.percentageValue
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import javax.imageio.ImageIO

/**
 * Start a new process that receive a image path with an image [inputImagePath] and copy and reduce the image in another
 * path [outputImagePath] using the give [percent]
 */
class ImageResizeTask (private var inputImagePath: String?,
                       private var outputImagePath: String?,
                       private var percent: Double) : Runnable {

    private val logger = KotlinLogging.logger {}

    override fun run() {
        try {
            resize(inputImagePath, outputImagePath, percent)
        } catch (e: IOException) {
            logger.error("Error starting thread", e)
        }
    }

    private fun resize(inputImagePath: String, outputImagePath: String, scaledWidth: Int, scaledHeight: Int) {
        val cleanedInputImagePath = URLDecoder.decode(inputImagePath, StandardCharsets.UTF_8.toString())
        val cleanedOutputImagePath = URLDecoder.decode(outputImagePath, StandardCharsets.UTF_8.toString())

        // reads input image
        val inputFile = File(cleanPath(cleanedInputImagePath))
        val inputImage = ImageIO.read(inputFile)

        // creates output image
        val outputImage = BufferedImage(scaledWidth, scaledHeight, inputImage.type)

        // scales the input image to the output image
        val g2d = outputImage.createGraphics()
        g2d.drawImage(inputImage, 0, 0, scaledWidth, scaledHeight, null)
        g2d.dispose()

        // extracts extension of output file
        val formatName = cleanedOutputImagePath.substring(cleanedOutputImagePath.lastIndexOf(".") + 1)

        // writes to output file
        val outputFile = File(cleanPath(cleanedOutputImagePath))
        if (!outputFile.exists()) {
            outputFile.createNewFile()
        }
        ImageIO.write(outputImage, formatName, outputFile)
    }

    private fun resize(inputImagePath: String?, outputImagePath: String?, percent: Double) {
        if (inputImagePath == null || inputImagePath == "") {
            return
        }
        if (outputImagePath == null || outputImagePath == "") {
            return
        }

        val cleanedInputImagePath  = URLDecoder.decode(inputImagePath, StandardCharsets.UTF_8.toString())
        val cleanedOutputImagePath = URLDecoder.decode(outputImagePath, StandardCharsets.UTF_8.toString())

        val file = File(cleanPath(cleanedInputImagePath))

        val inputImage: BufferedImage?  = ImageIO.read(file)

        if (inputImage != null) {
            val scaledWidth: Int = percentageValue(percent, inputImage.width.toDouble()).toInt()
            val scaledHeight: Int = percentageValue(percent, inputImage.height.toDouble()).toInt()

            resize(cleanedInputImagePath, cleanedOutputImagePath, scaledWidth, scaledHeight)
        }
    }
}